beacon_node/network/src/prefix_mapping.rs

```diff
@@ -0,0 +1,83 @@
+use lighthouse_network::discv5::enr::NodeId;
+use std::collections::HashMap;
+use types::{ChainSpec, Epoch, EthSpec, SubnetId};
+
+/// Stores mappings of `SubnetId` to `NodeId`s for prefix search.
+pub(crate) struct PrefixMapping {
+    chain_spec: ChainSpec,
+    epoch: Epoch,
+    mapping: HashMap<SubnetId, Vec<NodeId>>,
+}
+
+impl PrefixMapping {
+    pub fn new(spec: ChainSpec) -> Self {
+        Self {
+            chain_spec: spec,
+            epoch: Epoch::new(0),
+            mapping: HashMap::new(),
+        }
+    }
+
+    /// Returns `NodeId`s with the prefix that should be subscribed to the given `SubnetId` and `Epoch`.
+    pub fn get_target_nodes<TSpec: EthSpec>(
+        &mut self,
+        subnet_id: &SubnetId,
+        current_epoch: Epoch,
+    ) -> Result<Vec<NodeId>, &'static str> {
+        if self.epoch != current_epoch {
+            // compute prefixes
+            let computed_mapping = SubnetId::compute_prefix_mapping_for_epoch::<TSpec>(
+                current_epoch,
+                &self.chain_spec,
+            )?;
+
+            self.mapping = computed_mapping
+                .into_iter()
+                .map(|(subnet_id, ids)| {
+                    // convert `U256`s to `NodeId`s
+                    let node_ids = ids
+                        .into_iter()
+                        .map(|id| {
+                            let raw_node_id: [u8; 32] = id.into();
```

Comment:
The calculation for node-ids which comes from `compute_prefix_mapping_for_epoch` generates node-ids that contain the prefix, followed by 0's for the rest. i.e XXXX0000000000000000000000...

The problem with using this as a target for searching in discovery is that it is predictable and malicious actors can take advantage of discovery queries if they always know what the target is going to be. 

When we do the searchers, we should randomize the trailing bits, so that we get a unique set of peers inside each prefix. 

I think the best way, might be to just store the prefix bytes in this mapping. So that every epoch we calculate the new prefix bytes and delete the old ones. Then when we do the actual search, we generate a random target node-id and replace the first few bytes with the prefix we find here. This way, every time we do a search we have a random target.